source/brailleDisplayDrivers/freedomScientific.py

```diff
@@ -1,292 +1,546 @@
+FS_PKT_HVADJ = b"\x08"
+#: Lets the display beep. Not yet used in this driver.
+FS_PKT_BEEP = b"\x09"
+#: Sends a configuration request to the display. Mainly used to enable extended key mode on newer displays to use all the buttons.
+FS_PKT_CONFIG = b"\x0F"
+#: Indicates a response to FS_PKT_QUERY from the display
+FS_PKT_INFO = b"\x80"
+#: Sends braille cells to the display.
+FS_PKT_WRITE = b"\x81"
+#: Indicates extended keys have been pressed. Newer displays use this for some of their keys, see also the list in KeyGesture.extendedKeyLabels
+FS_PKT_EXT_KEY = b"\x82"
+
+# Parts of packets
+#: An empty packet argument or null byte
+FS_BYTE_NULL = b"\x00"
+#: Empty data in the packet payload
+FS_DATA_EMPTY = b""
+
+# FS_PKT_INFO payload offsets
+#: Start position of manufacturer in FS_PKT_INFO payload
+INFO_MANU_START = 0
+#: End position of manufacturer in FS_PKT_INFO payload
+INFO_MANU_END = 24
+#: Start position of model in FS_PKT_INFO payload
+INFO_MODEL_START = INFO_MANU_END
+#: End position of model in FS_PKT_INFO payload
+INFO_MODEL_END = INFO_MODEL_START + 16
+#: Start position of firmware version in FS_PKT_INFO payload
+INFO_VERSION_START = INFO_MODEL_END
+#: End position of firmware version in FS_PKT_INFO payload
+INFO_VERSION_END = INFO_MODEL_END + 8
+
+# Braille translation
+#: The number of dots in a braille character/cell
+DOTS_TABLE_SIZE = 8
+#: THe size of a full braille translation table including all possible dot combinations
+TRANSLATION_TABLE_SIZE = 2 ** DOTS_TABLE_SIZE
+
+def _makeTranslationTable(dotsTable):
+	"""Create a translation table for braille dot combinations
+
+	@param dotsTable: The list of 8 bitmasks to use for each dot (dot 1 - 8)
+	"""
+	def isoDot(number):
+		"""
+		Returns the ISO 11548 formatted braille dot for the given number.
+
+		From least- to most-significant octal digit:
+		
+		* the first contains dots 1-3
+		* the second contains dots 4-6
+		* the third contains dots 7-8
+
+		Based on: https://github.com/brltty/brltty/blob/master/Headers/brl_dots.h
+
+		@param number: The dot to encode (1-8)
+		@type number: int
+		"""
+		return 1 << (number - 1)
+
+	outputTable = [0] * TRANSLATION_TABLE_SIZE
+	for byte in xrange(TRANSLATION_TABLE_SIZE):
+		cell = 0
+		for dot in xrange(DOTS_TABLE_SIZE):
+			if byte & isoDot(dot + 1):
+				cell |= dotsTable[dot]
+		outputTable[byte] = cell
+	return outputTable
+
+def _translate(cells, translationTable):
+	"""Translate cells according to a translation table
+
+	The translation table contains the bytes to encode all the possible dot combinations.
+	See L{_makeTranslationTable} as well.
+
+	@param cells: The cells to translate, given in ISO 11548 format (used by most braille displays)
+	@type cells: [int]
+	@param translationTable: A list of all possible braille dot combinations
+	@type translationTable: [int]
+	"""
+	outCells = [0] * len(cells)
+	for i, cell in enumerate(cells):
+		outCells[i] = dotsTable[cell]

```

Comment:
Is this a bug, should `dotsTable` actually be `translationTable`

Reply:
I'm quite sure this is a bug. I have pushed a fix with many other small changes in https://github.com/bramd/nvda/pull/3